test/framework/src/main/java/org/elasticsearch/index/shard/IndexShardTestCase.java

```diff
@@ -0,0 +1,375 @@
+        closeShards(Arrays.asList(shards));
+    }
+
+    protected void closeShards(Iterable<IndexShard> shards) throws IOException {
+        for (IndexShard shard : shards) {
+            if (shard != null) {
+                shard.close("test", false);
+                IOUtils.close(shard.store());
+            }
+        }
+    }
+
+    protected void recoveryShardFromStore(IndexShard primary) throws IOException {
+        primary.markAsRecovering("store", new RecoveryState(primary.routingEntry(),
+            getFakeDiscoNode(primary.routingEntry().currentNodeId()),
+            null));
+        primary.recoverFromStore();
+        primary.updateRoutingEntry(ShardRoutingHelper.moveToStarted(primary.routingEntry()));
+    }
+
+    protected void recoveryEmptyReplica(IndexShard replica) throws IOException {
+        IndexShard primary = null;
+        try {
+            primary = newStartedShard(true);
+            recoverReplica(replica, primary);
+        } finally {
+            closeShards(primary);
+        }
+    }
+
+    private DiscoveryNode getFakeDiscoNode(String id) {
+        return new DiscoveryNode(id, new LocalTransportAddress("_fake_" + id), Version.CURRENT);
+    }
+
+    protected void recoverReplica(IndexShard replica, IndexShard primary) throws IOException {
+        recoverReplica(replica, primary,
+            (r, sourceNode) -> new RecoveryTarget(r, sourceNode, recoveryListener, version -> {}),
+            true);
+    }
+
+    protected void recoverReplica(IndexShard replica, IndexShard primary,
+                               BiFunction<IndexShard, DiscoveryNode, RecoveryTarget> targetSupplier,
+                               boolean markAsRecovering)
+        throws IOException {
+        final DiscoveryNode pNode = getFakeDiscoNode(primary.routingEntry().currentNodeId());
+        final DiscoveryNode rNode = getFakeDiscoNode(replica.routingEntry().currentNodeId());
+        if (markAsRecovering) {
+            replica.markAsRecovering("remote",
+                new RecoveryState(replica.routingEntry(), pNode, rNode));
+        } else {
+            assertEquals(replica.state(), IndexShardState.RECOVERING);
+        }
+        replica.prepareForIndexRecovery();
+        RecoveryTarget recoveryTarget = targetSupplier.apply(replica, pNode);
+        StartRecoveryRequest request = new StartRecoveryRequest(replica.shardId(), pNode, rNode,
+            getMetadataSnapshotOrEmpty(replica), false, 0);
+        RecoverySourceHandler recovery = new RecoverySourceHandler(primary, recoveryTarget, request, () -> 0L, e -> () -> {},
+            (int) ByteSizeUnit.MB.toKB(1), logger);
+        recovery.recoverToTarget();
+        recoveryTarget.markAsDone();
+        replica.updateRoutingEntry(ShardRoutingHelper.moveToStarted(replica.routingEntry()));
+    }
+
+    private Store.MetadataSnapshot getMetadataSnapshotOrEmpty(IndexShard replica) throws IOException {
+        Store.MetadataSnapshot result;
```

Comment:
compiler is not smart enough to see it's OK and complains about double assignment.